conversion_helpers.go

```diff
@@ -14,6 +14,45 @@ import (
 //   - conversion function from an external type to an internal type
 // ****************************************************************************
 
+// PayloadStringer can be used to convert payload bytes to string to show in web ui.
+type PayloadStringer interface {
+	String([]byte) string
+}
+
+type PayloadStringerFunc func([]byte) string
+
+func (f PayloadStringerFunc) String(b []byte) string {
+	return f(b)
+}
+
+var defaultPayloadStringer = PayloadStringerFunc(func(payload []byte) string {
+	if !isPrintable(payload) {
+		return "non-printable bytes"
+	}
+	return string(payload)
+})
+
+// isPrintable reports whether the given data is comprised of all printable runes.
+func isPrintable(data []byte) bool {
+	if !utf8.Valid(data) {
+		return false
+	}
+	isAllSpace := true
+	for _, r := range string(data) {
+		if !unicode.IsPrint(r) {
+			return false
+		}
+		if !unicode.IsSpace(r) {
+			isAllSpace = false
+		}
+	}
+	return !isAllSpace
+}
+
+type transformer struct {
```

Comment:
question: Why do we need this type?

Reply:
First I tried to have the `BytesStringer` at a package level global scope, and have a Set function. Later I thought, it'd be nice to pass it as an option while initialising the router. So I needed to have a BytesStringer reference ready to use in conversion helpers, I figured that doing it this way would require least possible passing around of BytesStringer.

If you have any other suggestion on how to go about it, let me know.

Reply:
Never-mind, I removed the type and passed the BytesStringer around.